src/include/serial_inline.h

```diff
@@ -79,13 +81,16 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Use an atomic load to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = WT_ATOMIC_LOAD(ins_stack[i], WT_ATOMIC_ACQUIRE);
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(
+            ins_stack[i], old_ins, new_ins, WT_ATOMIC_RELEASE, WT_ATOMIC_RELAXED))
             return (i == 0 ? WT_RESTART : 0);
-        if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
-            ins_head->tail[i] = new_ins;
+        tail = WT_ATOMIC_LOAD(&ins_head->tail[i], WT_ATOMIC_ACQUIRE);
+        if (tail == NULL || ins_stack[i] == &tail->next[i])
+            WT_ATOMIC_STORE(&ins_head->tail[i], new_ins, WT_ATOMIC_RELEASE);
```

Comment:
Staring at these lines a bit more, should this store be a CAS instead? What should happen if we raced with someone else storing a different pointer to `ins_head->tail[i]`? Is that impossible for some reason? Is it harmless? Is it a bug?